src/qtiCommonRenderer/renderers/interactions/ExtendedTextInteraction.js

```diff
@@ -811,6 +807,208 @@ function inputLimiter(interaction) {
     return limiter;
 }
 
+/**
+ * In Safari & `writing-mode: vertical-rl`, for *multiline* text:
+ *   - when user is typing, browser doesn't always repaint, so user doesn't see what he's typing.
+ *   - when user clicks (or key-navigates) inside text, cursor is shown in wrong position.
+ *     `textarea.selectionStart/textarea.selectionEnd` are set correctly, just visually cursor is painted in a wrong place.
+ * Create a workaround:
+ *   - force browser to repaint on every typed letter, using a trick
+ *   - hide native cursor, show custom div as a cursor instead
+ *   - position this custom cursor:
+ *     - create a shadow div contianing same text as textarea (<textarea>ab</textarea> -> <shadow><span>a</span><span>b</span></shadow>)
+ *     - shadow div should also have same size, font, and text-wrapping rules.
+ *     - when after click/key-nav browser changes `textarea.selectionStart` to N, use shadow to understand what are the coordinates of the Nth letter of text
+ *     - then draw custom cursor at those coordinates.
+ * Unsolved issues:
+ *   - when user clicks between lines (in a space left by the difference of line-height & font-size), `textarea.selectionStart` is wrong, cursor jumps randomly
+ *   - when user clicks on empty space left in the line, or at empty lines, `textarea.selectionStart` is wrong, cursor jumps randomly
+ *   - key-navigation is not intuitive (left/right still jump letters, and up/down - lines, as in horizontal-tb)
+ * @param {JQuery} $textarea
+ * @param {string} serial
+ * @returns {Object|null} if safari & vertical-rl,`{ syncValue: () => void, destroy: () => void }` . Otherwise `null`.
+ */
+function _patchSafariVerticalRl($textarea, serial) {
+    if (!getIsItemWritingModeVerticalRl() || !isSafari()) {
+        return null;
+    }
+
+    $textarea.addClass('hide-caret');
+
+    const $shadow = $('<div>', { class: 'shadow-textarea' });
+    const $cursor = $('<div>', { class: 'extendedText-shadow-caret' });
+    $(document.body).append($cursor);
+    $textarea.after($('<div>', { class: 'shadow-container' }).append($shadow));
+
+    let repaintIdx = 0;
+
+    function setShadowString(str) {
+        const shadowString = [...str, ' ']
+            .map(i => {
+                if (i === '\n') {
+                    return '<br/>';
+                } else if (i === ' ') {
+                    return `<span>&#x20;</span>`;
+                } else {
+                    return `<span>${i}</span>`;
+                }
+            })
+            .join('');
+
+        $shadow.get(0).innerHTML = shadowString;
+    }
+
+    function forceTextareaRepaint() {
+        if (repaintIdx % 2 === 0) {
+            $textarea.get(0).style.opacity = '99%';
+        } else {
+            $textarea.get(0).style.opacity = '98%';
+        }
```

Comment:
Don't know if you have investigated [will-change](https://developer.mozilla.org/en-US/docs/Web/CSS/will-change#examples) in anticipation of these opacity changes?

Reply:
No, it's an interesting idea.
But it will optimize rendering, and we here **do not need** it to be optimized. Because we are trying to force browser to repaint more often. And if `will-change` turns on some kind of caching, it may create another issue with repainting?

PS: Seems to be the case in browserstack Safari 18.4 at least:

https://github.com/user-attachments/assets/586bd8f5-c2c2-4882-a85a-639a77bc4194